logical/framework/openapi.go

```diff
@@ -0,0 +1,548 @@
+		if sp == path ||
+			(strings.HasSuffix(sp, "*") && strings.HasPrefix(path, sp[0:len(sp)-1])) {
+			return true
+		}
+	}
+	return false
+}
+
+// expandPattern expands a regex pattern by generating permutations of any optional parameters
+// and changing named parameters into their {openapi} equivalents.
+func expandPattern(pattern string) []string {
+	var paths []string
+
+	// GenericNameRegex adds a regex that complicates our parsing. It is much easier to
+	// detect and remove it now than to compensate for in the other regexes.
+	//
+	// example: (?P<foo>\\w(([\\w-.]+)?\\w)?) -> (?P<foo>)
+	base := GenericNameRegex("")
+	start := strings.Index(base, ">")
+	end := strings.LastIndex(base, ")")
+	regexToRemove := ""
+	if start != -1 && end != -1 && end > start {
+		regexToRemove = base[start+1 : end]
+	}
+
+	pattern = strings.Replace(pattern, regexToRemove, "", -1)
+
+	// Initialize paths with the original pattern or the halves of an
+	// alternation, which is also present in some patterns.
+	matches := altRe.FindAllStringSubmatch(pattern, -1)
+	if len(matches) > 0 {
+		paths = []string{matches[0][1], matches[0][2]}
+	} else {
+		paths = []string{pattern}
+	}
+
+	// Expand all optional regex elements into two paths. This approach is really only useful up to 2 optional
+	// groups, but we probably don't want to deal with the exponential increase beyond that anyway.
+	for i := 0; i < len(paths); i++ {
+		p := paths[i]
+
+		// match is a 2-element slice that will have a start and end index
+		// for the left-most match of a regex of form: (lease/)?
+		match := optRe.FindStringIndex(p)
+
+		if match != nil {
+			// create a path that includes the optional element but without
+			// parenthesis or the '?' character.
+			paths[i] = p[:match[0]] + p[match[0]+1:match[1]-2] + p[match[1]:]
+
+			// create a path that excludes the optional element.
+			paths = append(paths, p[:match[0]]+p[match[1]:])
+			i--
+		}
+	}
+
+	// Replace named parameters (?P<foo>) with {foo}
+	var replacedPaths []string
+
+	for _, path := range paths {
+		result := reqdRe.FindAllStringSubmatch(path, -1)
+		if result != nil {
+			for _, p := range result {
+				par := p[1]
+				path = strings.Replace(path, p[0], fmt.Sprintf("{%s}", par), 1)
+			}
+		}
+		// Final cleanup
+		path = cleanSuffixRe.ReplaceAllString(path, "")
+		path = cleanCharsRe.ReplaceAllString(path, "")
+		replacedPaths = append(replacedPaths, path)
+	}
+
+	return replacedPaths
+}
+
+// schemaType is a subset of the JSON Schema elements used as a target
+// for conversions from Vault's standard FieldTypes.
+type schemaType struct {
+	baseType string
+	items    string
+	format   string
+}
+
+// convertType translates a FieldType into an OpenAPI type.
+// In the case of arrays, a subtype is returned as well.
+func convertType(t FieldType) schemaType {
+	ret := schemaType{}
+
+	switch t {
+	case TypeString, TypeNameString, TypeLowerCaseString, TypeHeader:
+		ret.baseType = "string"
```

Comment:
Is there any way to better indicate to the user when we e.g. lowercase a value for them?

Reply:
Yep. `format` isn't strongly specified and can be "lowercase" (baseType is still string).

Reply:
Sure, but -- these don't have a `format` parameter. Perhaps they should?

Reply:
I did make a change to give TypeLowerCaseString `format = "lowercase"`. TypeHeader will already get special designation as being located in the header.  There is another supported field that may be worth adding: `pattern`. This is a regex. The TypeNameString regex could be put in there, for example.

Reply:
I've added the "pattern" field and updated TypeNameString to populate it with that regex.

Reply:
Thanks!